drake/solvers/Constraint.h

```diff
@@ -193,6 +195,48 @@ class BoundingBoxConstraint : public LinearConstraint {
     y = x;
   }
 };
+
+
+/** LinearComplementarityConstraint
+ *
+ * Implements a constraint of the form:
+ *   Mx + q >= 0
+ *   x >= 0
+ *   x'(Mx + q) == 0
+ *
+ * An implied slack variable complements any 0 component of x.  To get
+ * the slack values at a given solution x, use eval(x).
+ */
+class LinearComplementarityConstraint : public Constraint {
+ public:
+  template <typename DerivedM, typename Derivedq>
+  LinearComplementarityConstraint(const Eigen::MatrixBase<DerivedM>& M,
+                                  const Eigen::MatrixBase<Derivedq>& q)
+      : Constraint(q.rows()), M(M), q(q) {}
+
+  virtual ~LinearComplementarityConstraint() {}
+
+  /** Return Mx + q (the value of the slack variable). */
+  virtual void eval(const Eigen::Ref<const Eigen::VectorXd>& x,
+                    Eigen::VectorXd& y) const override {
+    y.resize(getNumConstraints());
+    y = (M * x) + q;
+  }
+  virtual void eval(const Eigen::Ref<const TaylorVecXd>& x,
+                    TaylorVecXd& y) const override {
+    y.resize(getNumConstraints());
+    y = (M.cast<TaylorVarXd>() * x) + q.cast<TaylorVarXd>();
+  };
+
+  const Eigen::MatrixXd& getM() const { return M; };
+  const Eigen::VectorXd& getq() const { return q; };
+
+ private:
+  // TODO ggould We are storing what are likely statically sized matrices
+  // in dynamically allocated containers.  This probably isn't optimal.
+  Eigen::MatrixXd M;
```

Comment:
@david-german-tri :
While I agree those must come up to spec, the function names are consistent with the superclass and with the convention already present in the file, so the whole file (and its callers) needs to be updated; I don't think this PR is the place for that.  I will prepare a separate PR for bringing function names up to spec.

The variable names correspond to the standard usage in the literature (capital letters for matrices, lowercase for vectors and scalars) and I believe that it would be bad to lose the clarity that provides.  In any event this issue/PR is probably not the place for that discussion; we can continue it offline or open a new issue.